### PR TITLE
[ci] Save built wheel when test fails

### PR DESCRIPTION
**Pull request by feisuzhu**

### CODE REVIEW

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -171,6 +172,13 @@ jobs:
         run: |
           docker rm taichi_build taichi_test -f
 
+      - name: Save wheel if test failed
+        if: failure() && steps.test.conclusion == 'failure'
```
Comment (qiao-bo) [inline .github/workflows/testing.yml:176]: Seems the github syntax says `${{ failure() && steps.demo.conclusion == 'failure' }}` ? https://docs.github.com/en/actions/learn-github-actions/expressions#example-10
are these the same? ;/

Comment (feisuzhu) [reply .github/workflows/testing.yml:176]: It says

> When you use expressions in an if conditional, you may omit the expression syntax 

in the same page :)

Comment (qiao-bo) [reply .github/workflows/testing.yml:176]: cool, didn't know that

**Review by qiao-bo — APPROVED** [review]
LGTM!

File: .github/workflows/testing.yml (hunk 1 above)
Comment (frostming) [reply .github/workflows/testing.yml:176]: What is the difference between `failure()` and `steps.demo.conclusion == 'failure'`?

Comment (qiao-bo) [reply .github/workflows/testing.yml:176]: > What is the difference between `failure()` and `steps.demo.conclusion == 'failure'`? Or does one imply the other?

as far as i understand from the doc (link above), the first part is true as long as any previous steps fail, while the second part specifies a condition for that failure, namely `steps.test.conclusion` the test (id) part should fail, not the build step for example.

Comment (frostming) [reply .github/workflows/testing.yml:176]: Fine, as the example uses that expression.

**Pull request merged**